include/core/CMemory.h

```diff
@@ -697,26 +700,27 @@ class CORE_EXPORT CMemoryDebug : private CNonInstantiatable {
     static void dynamicSize(const char* name,
                             const std::shared_ptr<T>& t,
                             CMemoryUsage::TMemoryUsagePtr mem) {
-        if (t != nullptr) {
-            long uc = t.use_count();
-            // If the pointer is shared by multiple users, each one
-            // might count it, so divide by the number of users.
-            // However, if only 1 user has it, do a full debug.
-            if (uc == 1) {
-                // Note we add on sizeof(long) here to account for
-                // the memory used by the shared reference count.
-                mem->addItem("shared_ptr", sizeof(long) + CMemory::staticSize(*t));
-                dynamicSize(name, *t, mem);
-            } else {
-                std::ostringstream ss;
-                ss << "shared_ptr (x" << uc << ')';
-                // Note we add on sizeof(long) here to account for
-                // the memory used by the shared reference count.
-                // Also, round up.
-                mem->addItem(ss.str(), (sizeof(long) + CMemory::staticSize(*t) +
-                                        CMemory::dynamicSize(*t) + std::size_t(uc - 1)) /
-                                           uc);
-            }
+        if (t == nullptr || t.use_count() == 0) {
+            return;
+        }
+        long uc = t.use_count();
```

Comment:
Since `use_count` is 0 when the pointer is null, maybe we could _just_ check this? Same in the non-debug version of the method.

Also, this code is reading `use_count` twice and the value could change between the two calls. So I think it’s best to cache its value in `uc` as the first step.

Reply:
Weirdly I've just encountered a case where `use_count` returns 1 for a `nullptr`

```
Exception = EXC_BAD_ACCESS (code=1, address=0x0)
t = {const std::__1::shared_ptr<ml::maths::CPrior> &} nullptr
 __ptr_ = {std::__1::shared_ptr<ml::maths::CPrior>::element_type * | 0x0} NULL
uc = {long} 1
```

I _think_ this may be related to https://stackoverflow.com/questions/48885252/c-sharedptr-use-count-for-nullptr/48885643. So an explicit check for `nullptr` is still needed to trap this case

Reply:
Interesting, thanks for finding that link.  It would still be good to only read `use_count` once, but, like you say, we'll have to keep the `nullptr` check.